### PR TITLE
IntegerLiteralExpr now is lowered directly into SIL.

### PR DESCRIPTION
**Pull request by pschuh**
For context, String, Nil, and Bool already behave this way.

Note: Before it used to construct (call, ... (integer_literal)), and the
call would be made explicit / implicit based on if you did eg: Int(3) or
just 3. This however did not translate to the new world so this PR adds
a IsExplicitConversion bit to NumberLiteralExpr which captures this lost info. 
Some side results of all this are that some warnings changed a little and some
instructions are emitted in a different order.

### CODE REVIEW

File: lib/SILGen/SILGenApply.cpp [hunk 1]
```diff
@@ -5565,8 +5575,35 @@ RValue SILGenFunction::emitLiteral(LiteralExpr *literal, SGFContext C) {
...
+        Value = ctx.SourceMgr.getLineAndColumn(Loc).second;
+
+      auto Ty = SILType::getBuiltinIntegerLiteralType(ctx);
+      SILValue V = B.createIntegerLiteral(literal, Ty, Value);
+      builtinLiteralArgs = RValue(*this, {ManagedValue::forUnmanaged(V)},
+                                  Ty.getASTType()->getCanonicalType());
+      builtinInit = magicLiteral->getBuiltinInitializer();
+      init = magicLiteral->getInitializer();
+      break;
+    }
```
Comment (DougGregor) [inline lib/SILGen/SILGenApply.cpp]: There is a bunch of duplication here. Can you collapse the two cases together and use the kind to decide between the `first` and `second` of `getLineAndColumn`?

**Review by DougGregor — APPROVED** [review]
This is so cool, thank you!

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Review by slavapestov — APPROVED** [review]
Awesome!

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by CodaFi** [comment]
Do this to floats and we can flush `convertLiteral` entirely!

**Comment by pschuh** [comment]
@swift-ci Please smoke test

**Comment by pschuh** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@CodaFi that's the plan. @pschuh has been systemically working through all the cases. After convertLiteral() is gone, we can do two more things:

- Change ClangImporter to synthesize fully-type checked AST for constants. This will speed up type checking in non-WMO mode (since we end up importing the same constants over and over again for each frontend job).

- Pick off more uses of callWitness(). convertLiteral() was one of the remaining usages.

**Comment by brentdax** [comment]
> Pick off more uses of callWitness(), using the same technique of storing a ConcreteDeclRef and emitting the call in SILGen instead of building CallExprs in the type checker. convertLiteral() was one of the remaining usages of callWitness(). The others include:
>   * String interpolation

Slava just asked privately, so I'll repeat here that you should be able to remove string interpolation's use of `callWitness()` without regressing any intended overload-handling behavior. The string interpolation calls which are supposed to evaluate overloads are built in ParseExpr.cpp using `UnresolvedDotExpr`; the CSApply.cpp uses of `callWitness()` should not select overloads.

**Pull request merged**